Add kubernetes labels

dhpollack opened the pull request:
This adds support for adding kubernetes labels to pods.  Either via the kubernetes decorator with `labels={"label": "value", "label2": "value2"}`, with an env var `METAFLOW_KUBERNETES_LABELS="label=value,label2=value2"`, or add to the with command `python hello_world.py run --with kubernetes:labels="cli_label=cli_value"`.  I also created an issue for this:

Netflix/metaflow#1235

dhpollack commented on metaflow/plugins/kubernetes/kubernetes_cli.py:
```diff
@@ -175,7 +182,13 @@ def echo(msg, stream="stderr", job_id=None):
+
+    # `labels` is a tuple of strings or a tuple with a single comma separated string
+    # convert it to a dict
+    labels = KubernetesDecorator.parse_kube_list(
```
The string generated in the `runtime_step_cli` in the `kubernetes_decorator.py` ends up looking something like:

```
labels = ("label1=val1,label2=val2",)
```

This seems somewhat by design for the `node_selector` field.  Hence there is a bit of additional logic here.

dhpollack replied:
https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/kubernetes/kubernetes_decorator.py#L288

I suspect this should not have `",".join(...)`.  I will test if removing that fixes the issue.

odracci commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -103,9 +109,16 @@ def __init__(self, attributes=None, statically_defined=False):
+            self.attributes["labels"] = KUBERNETES_LABELS
+
+        if isinstance(self.attributes["labels"], str):
+            self.attributes["labels"] = self.parse_kube_list(
+                self.attributes["labels"].split(","), False
+            )
```
I would validate the labels, using something like https://github.com/apache/airflow/blob/207f65b542a8aa212f04a9d252762643cfd67a74/airflow/kubernetes/pod_generator.py#L56-L74
[Kubernetes documentation](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)

dhpollack replied:
done

dhpollack commented:
> Hey @dhpollack, thanks for looking into this. Metaflow also has the concept of `tags` (https://docs.metaflow.org/scaling/tagging).
> 
> Is there any overlap between `tags` and labels? And could there be some option to simply use one to avoid confusion? WDYT?

We use kubernetes labels everywhere in our organization for cost tracking and tagging. Some of these labels are not very useful as metaflow tags.

dhpollack commented:
> Hey @dhpollack, thanks for looking into this. Metaflow also has the concept of `tags` (https://docs.metaflow.org/scaling/tagging).
> 
> Is there any overlap between `tags` and labels? And could there be some option to simply use one to avoid confusion? WDYT?

But specifically about overlap... There is a little but they are not the same nor do I think one should try to use them in the same. 

For one, your tags are strings, while the kubernetes labels are key-valur pairs. But even from a user perspective, metaflow tags seem designed for data scientists to track experiments in metaflow. While kubernetes labels are what a lot of the kubernetes universe uses for filtering and organizing things that happen on kubernetes...

If you changed metaflow tags to mirror kubernetes labels (key-valur, restrictions on length and valid characters, etc etc) then i think you could simply use one of them. But that's a lot more work for you and I think people that want to add kubernetes labels to their pods generally know what kubernetes labels are and wouldn't be too bothered by the additional complexity.

dhpollack commented:
@shrinandj can you re-run the failed CI test?  It looks like it failed but shouldn't have.  Probably just a hiccup on the machine.

shrinandj commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -384,13 +401,34 @@ def _save_package_once(cls, flow_datastore, package):
 
     @staticmethod
-    def parse_node_selector(node_selector: list):
+    def parse_kube_list(items: Union[str, List[str]], requires_both: bool = True):
```
Is the name `parse_kube_list` seem confusing? Would something like `parse_kube_keyvalue_list` be better? It also makes the function argument `requires_both` more intuitive.

WDYT?

dhpollack replied:
cool, I made this change.

shrinandj commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -384,13 +401,34 @@ def _save_package_once(cls, flow_datastore, package):
+    def clean_kube_labels(
+        labels: Optional[Dict[str, Optional[str]]],
+        max_len: int = 63,
+        regex_sub: str = r"^[^a-z0-9A-Z]*|[^a-zA-Z0-9_\-\.]|[^a-z0-9A-Z]*$",
```
Can you confirm that this was tested thoroughly?

dhpollack replied:
Ok, I added some tests for this function.  I also added an assertion error if the label is entirely composed of non-valid characters.  This seems like an edge case that would fail while launching the job.

shrinandj commented:
Generally looks good. Thanks for doing this @dhpollack! Can you add some details of the tests that were run for testing this PR? Especially, did you create argo workflows with this code and ensure that the pods created by Argo have the correct labels?

dhpollack commented:
So to test this I used something like the following flow:

```python
from metaflow import FlowSpec, step, kubernetes

class TestFlow(FlowSpec):
    @step
    def start(self):
        self.my_var = "hello world"
        self.next(self.a)

    @kubernetes(labels={"my_label": "my_value", "label2": None, "label3": "val"})
    @step
    def a(self):
        print("the data artifact is: %s" % self.my_var)
        self.next(self.b)

    @kubernetes(
        labels={
            "my_label": "superlonglabelthatshouldgethashedbytheautomatedhashingfunctionorsomethinglikethat",  # noqa: E501
            "my_label2": "ein bißchen falsch",
        }
    )
    @step
    def b(self):
        pass
        self.next(self.end)

    @step
    def end(self):
        print("the data artifact is still: %s" % self.my_var)

if __name__ == "__main__":
    TestFlow()
```

Then I ran the following (note image each space is a new terminal):

```bash
export METAFLOW_KUBERNETES_LABELS='env_label=env_var`
python test_workflow.py run  --with kubernetes

python test_workflow.py run

python test_workflow.py run --with kubernetes:labels="label1=val1,label2=val2,label3,label4=val4"

# remove kubernetes decorators from the flow (this is actually how we generally would use this)
export METAFLOW_KUBERNETES_LABELS='env_label=env_var`
python test_workflow.py run --with kubernetes

export METAFLOW_KUBERNETES_LABELS='env_label=env_var`
python test_workflow.py argo-workflows create && python test_workflows argo-workflows trigger
```

There are a few quirks.

1. If you set an env var and use `--with kubernetes` the steps decorated with the kubernetes labels overwrite the env var labels completely.
2. The argo labels seem to work, but it feels a bit like magic because metaflow automatically imports the kubernetes decorator and that adds the labels.  If you look at the argo template that gets generated, the labels are not in the argo metadata.  But I feel like that was ok, because if you wanted to set those then perhaps you'd have a separate parameter for the argo decorator that is a flow decorator rather than a step decorator.

shrinandj commented on test/unit/test_kubernetes_decorator.py:
```diff
@@ -0,0 +1,33 @@
+from metaflow.plugins.kubernetes.kubernetes_decorator import KubernetesDecorator
+
+
+@pytest.mark.parametrize(
```
This is awesome! Thank you so much for doing this!

shrinandj commented:
Thanks for the details about testing. It definitely increases the reviewers confidence in the commits.

Just a couple of other tests:
- Can you test with K8s labels AND nodeSelectors since this code touches that functionality as well.
- Can you test original `@kubernetes` and `--with kubernetes` functionality without labels. Basically, to ensure that nothing has regressed because of this change.

dhpollack commented:
> Thanks for the details about testing. It definitely increases the reviewers confidence in the commits.
> 
> Just a couple of other tests:
> 
>     * Can you test with K8s labels AND nodeSelectors since this code touches that functionality as well.
> 
>     * Can you test original `@kubernetes` and `--with kubernetes` functionality without labels. Basically, to ensure that nothing has regressed because of this change.

I forgot to mention that in my comment about the tests.  I also tested without labels and with node selectors in different combinations as well (alone, via the decorator, env vars, and using `--with kubernetes:...`.  The main change was the change in the [`runtime_step_cli`](https://github.com/Netflix/metaflow/pull/1236/files#diff-61935adb175a915b5065b6d681245a71e224de77f911687c1ffa3f48d5542aeaR304).  Previously, if you had more than one node selector, then it would create a tuple with a single item like `('label1=val1,label2=val2',)` and now it will do `('label1=val1', 'label2=val2')` which I think was just a bug in the previous code.

dhpollack commented:
@savingoyal Anything you would like done here?

shrinandj commented:
I download the source code locally and tried to test this change.

* Ran a flow with Kubernetes and a single label. This worked just fine.
```
$ python3 /tmp/a.py run --with kubernetes:labels="abc=pqr"
...
```

* Ran a flow with Kubernetes decorator and added a label there. This worked just fine.
```
    @kubernetes(labels={"abc":"pqr"})
```

* Ran a flow with Kubernetes decorator and add two labels. This worked just fine too.
```
    @kubernetes(labels={"abc":"pqr", "lmn":"xyz"})
```

* Ran a flow with Kubernetes on cli with two labels. **THIS FAILED** with the following error
```
$ python3 a.py run --with kubernetes:labels="abc=pqr,lmn=xyz"
Metaflow 2.7.19 executing StarterFlow for user:shri
    Unknown decorator attribute:
    Decorator 'kubernetes' does not support the attribute 'lmn'. These attributes are supported: cpu, memory, disk, image, service_account, secrets, node_selector, labels, namespace, gpu, gpu_vendor, tolerations.
```

* exported the env variable METAFLOW_KUBERNETES_LABELS with one label and ran a flow. This worked just fine.

* exported the env variable METAFLOW_KUBERNETES_LABELS with two labels and ran a flow. This worked just fine too.

* Ran a few other flows with Kubernetes specifying cpu, memory, etc. on the cli. They worked just fine.

dhpollack commented:
@shrinandj looking at how metaflow creates the `--with kubernetes` string for argo.  It actually does something like `--with kubernetes:labels='{"abc":"pqr", "lmn":"xyz"}'`.  I am not sure if this is undesired behavior or bad documentation.  The way that I had it before was based on the node selector, but that also did not work correctly when having multiple node selectors.

https://github.com/Netflix/metaflow/pull/1236#discussion_r1071193731

It's related to this issue which I brought up earlier.

dhpollack commented:
The following code will split `"labels=l1=v1,l2=v2"` as `[("labels", "l1=v1"), ("l2", "v2")]`.  It looks like the only way to have multiple objects in this format is to use a json object.  The other way would be to write a much more complicated regex that searches for valid keys and then rewrites that string as a json.

https://github.com/Netflix/metaflow/blob/master/metaflow/decorators.py#L132

```
    @classmethod
    def _parse_decorator_spec(cls, deco_spec):
        if len(deco_spec) == 0:
            return cls()

        attrs = {}
        # TODO: Do we really want to allow spaces in the names of attributes?!?
        for a in re.split(""",(?=[\s\w]+=)""", deco_spec):
            name, val = a.split("=", 1)
            ...
```

shrinandj commented:
There's some quirkiness involved related to the shell and the OS involved when there are multiple labels to be specified on the CLI. Using the json style input works just fine in that case.

Verified this with:
```
python a.py run --with kubernetes:labels='{"l1":"v1","l2":"v2"}'
```

This change should be good to go!

dhpollack commented on metaflow/plugins/kubernetes/kubernetes.py:
```diff
@@ -9,6 +9,8 @@
 from metaflow.metaflow_config import (
     SERVICE_HEADERS,
     SERVICE_INTERNAL_URL,
+    CARD_AZUREROOT,
```
should this be removed?

dhpollack commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -1,8 +1,10 @@
 import platform
-import requests
+import re
 import sys
```
```suggestion
import requests
```

savingoyal commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -399,14 +414,61 @@ def _save_package_once(cls, flow_datastore, package):
+            raise KubernetesException("Unable to parse kubernetes list: %s" % items)
+
+    @staticmethod
+    def clean_kube_labels(
```
@dhpollack what's the intention behind this? Ideally if we are supporting labels we would want to pass them as is to kubernetes.

savingoyal replied:
Mutating labels may catch the user off guard when the desired label isn't attached as is - we could rely on runtime checks disallowing certain labels instead - I agree with you that that will lead to an error further down during runtime but will be a better no-surprises approach. Also, we do expect that OPA may intercept the request and not allow certain kinds of labels be attached to the pod - so there is that use case. Would it be useful to jump on a quick 15-30 min call to quickly go over all open questions so that we can push this PR out asap? Let me know if any time [here](https://calendar.app.google/7wjKcswSBVdMgMbNA) works.

dhpollack replied:
I have changed this to throw an exception now.

saikonen commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -387,14 +403,59 @@ def _save_package_once(cls, flow_datastore, package):
+                return s
+            if not re.search(regex_match, s):
+                # this is the same message kubernetes itself returns
+                raise Exception(
```
use the MetaflowException wrapper to get a cleaner error output for the CLI

dhpollack replied:
Gonna use the `KubernetesException` which is a child of `MetaflowException`

saikonen commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -98,9 +105,17 @@ def __init__(self, attributes=None, statically_defined=False):
+            self.attributes["labels"] = self.parse_kube_keyvalue_list(
+                self.attributes["labels"].split(","), False
+            )
+        self.attributes["labels"] = self.validate_kube_labels(self.attributes["labels"])
```
Nit: relies on `validate_kube_labels` to return something, which given its name is unexpected. As the validator is more concerned about raising exceptions, it doesn't necessarily need to return anything. Could we instead simply call it to validate the labels in the attributes?

Based on the [lifecycle graph](https://github.com/Netflix/metaflow/blob/master/docs/lifecycle.png) , one place to call this could be in the `step_init()` phase

not a dealbreaker though so feel free to ignore this, as the current implementation also works :)

dhpollack replied:
So I changed the function to return a bool or throw an exception.

dhpollack commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -387,14 +403,59 @@ def _save_package_once(cls, flow_datastore, package):
+                )
+            return s
+
+        return {k: validate_label(v) for k, v in labels.items()} if labels else labels
```
The keys would need to pass a more complicated regex.  Basically, you can have a prefix for keys that is a valid DSN and can be up to 256 characters.  But also, one reason I am not validating the keys is that in theory the key is more static than the value.  I would just let bad keys fail and let the user get the kubernetes message.  If you want me to look for those reserved prefixes (`kubernetes.io` and `k8s.io`), I could easily do that.  I don't think it is worth it tho.

saikonen replied:
Agree that it might become unnecessarily complex with the key validation, we can skip this for the time being then. Maybe good to write a comment in the code for future developers as to why we only validate the values.

dhpollack commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -387,14 +403,59 @@ def _save_package_once(cls, flow_datastore, package):
+                    f'Invalid value: "{s}": a valid label must be an empty string or '
+                    "consist of alphanumeric characters, '-', '_' or '.', and must "
+                    "start and end with an alphanumeric character (e.g. 'MyValue', "
+                    "or 'my_value', or '12345', regex used for validation is "
+                    "'^(([A-Za-z0-9][-A-Za-z0-9_.]{0,61})?[A-Za-z0-9])?$'"
```
This is like the exact message that the kubernetes library gives you.  Would you prefer something like this:

```
                raise MetaflowException(
                    f'Invalid value: "{s}": a valid label must be an empty string or '
                    "consist of alphanumeric characters, '-', '_' or '.', must "
                    "start and end with an alphanumeric character (e.g. 'MyValue', "
                    "or 'my_value', or '12345'), and be no longer than 63 characters"
                )
```

saikonen replied:
It's a bit easier on the eyes without the regex in my opinion. I would maybe consider splitting the conditions into a list like
```
raise KubernetesException(
                    f'Invalid value: "{s}"\n'
                    "A valid label is an empty string or one that\n"
                    "   - Consists of alphanumeric, '-', '_', or '.' characters\n"
                    "   - Begins and ends with an alphanumeric character\n"
                    "   - Is at most 63 characters long"
                )
```
but the current change is good as well.

Note that there's still the comment about the exception being identical to what kubernetes throws which should be cleaned up.

saikonen commented on metaflow/plugins/kubernetes/kubernetes_decorator.py:
```diff
@@ -391,14 +407,80 @@ def _save_package_once(cls, flow_datastore, package):
+                if requires_both:
+                    item[1]  # raise IndexError
+                if str(item[0]) in ret:
+                    raise KubernetesException("Duplicate key found: %s" % str(item[0]))
```
In its current form the exception does not  convey where the duplicate key was found, as it makes no mention of 'labels' when I set duplicate labels.

As this is used for checks other than the labels as well, A more suitable place to perform the duplicate label check might be in the `validate_kube_labels` instead. That way the exception can be more specific as well (Duplicate *label* found)

dhpollack replied:
This function is used on both labels and node selectors and putting the check here ensures that we catch duplicate keys in both sets.

savingoyal commented:
@dhpollack looks like this PR is backward incompatible - 

<img width="805" alt="Screen Shot 2023-04-17 at 3 32 17 PM" src="https://user-images.githubusercontent.com/763451/232624647-7d3a5c22-c7b4-43c7-843e-6e1dfd168237.png">

I have created #1359 which reverts this PR so that releases are unblocked. Can you please address this issue and resubmit the PR?

dhpollack commented:
> @dhpollack looks like this PR is backward incompatible -
> 
> <img alt="Screen Shot 2023-04-17 at 3 32 17 PM" width="805" src="https://user-images.githubusercontent.com/763451/232624647-7d3a5c22-c7b4-43c7-843e-6e1dfd168237.png">
> 
> I have created #1359 which reverts this PR so that releases are unblocked. Can you please address this issue and resubmit the PR?

Yea, looks like it should be a quick fix.  I'll work on it tomorrow.
